### DISCUSSION TITLE
Can a standard compiler plugin add a main entry point?

### DISCUSSION THREAD

**Discussion by soronpo** [scala/scala3#20552] [Compiler internals Q&A]
I'm fiddling with a compiler plugin that adds a main entry point for companion objects of classes annotated with a certain annotation. 
Since I'm running into trouble, it's not clear if this is even possible to do with a standard plugin, so I'll be happy to even just get a clarification it's possible. Specifically:
1. If the companion object exists, adding a `def main(args: Array[String]): Unit` method to its body should be recognized.
2. It's possible to create a companion object (using `ModuleDef`) with a `def main(args: Array[String]): Unit` as its body for it to be recognized during runtime.

**Comment by soronpo** [comment]
Ok, point 1 is working.
Still need to check for point 2.

**Comment by soronpo** [comment]
I can't get 2 to work. Is creating a companion object allowed in a standard compiler plugin?

**Reply by sjrd** [reply]
The question is not so much whether it's *allowed* as whether it's *possible*. It should be possible, but there is almost certainly more wiring to do than "just" creating the module class and the module val. You may want to look at an existing post-typer phase of the compiler that synthesizes companion objects, to see how it does it. [`ExtensionMethods`](https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/dotc/transform/ExtensionMethods.scala) might be a good candidate.

**Reply by soronpo** [reply]
Thank you @sjrd for the pointer! I got it to almost work the way I want it.
There are two minor issues that maybe you can answer again if it's just possible to fix:
A. When I add the companion, the added companion module class now collides with the existing class that I create a companion for.
```
[warn] 6 |class Example() extends DFDesign
[warn]   |      ^
[warn]   |      Example and Example produce classes that overwrite one another
```
This is of course different behavior than when I wrote the companion manually, although both create another class in the same name. So I'm guessing I'm missing some wiring that explicitly defines the newly created class as a companion. Do you think it's possible to do in a post-typer plugin?
B. I now noticed that when I add the main entry point I can run it with `sbt runMain`, but `sbt run` does not detect it. Is this fixable? I'm not familiar between the two mechanisms.

**Reply by soronpo** [reply]
> B. I now noticed that when I add the main entry point I can run it with sbt runMain, but sbt run does not detect it. Is this fixable? I'm not familiar between the two mechanisms.

I managed to fix this by extending from `App`. It should be good enough.

**Reply by soronpo** [reply]
OK, extending `App` is not a good solution if I want argument support. Any clue why `sbt run` does not identify this main but `sbt runMain` does?

**Reply by sjrd** [reply]
`sbt run` takes its information from the Zinc analysis. So it's possible that the transformation comes too late for it to be registered in the sbt API generation phase.

**Reply by soronpo** [reply]
It seems if I publish locally and run sbt or scala-cli as usual it works fine. Don't know why, but I'll take the win.
Thank you very much!
